src/coreclr/jit/codegenxarch.cpp

```diff
@@ -5691,9 +5702,12 @@ void CodeGen::genCall(GenTreeCall* call)
     // To limit code size increase impact: we only issue VZEROUPPER before PInvoke call, not issue
     // VZEROUPPER after PInvoke call because transition penalty from legacy SSE to AVX only happens
     // when there's preceding 256-bit AVX to legacy SSE transition penalty.
-    if (call->IsPInvoke() && (call->gtCallType == CT_USER_FUNC) && GetEmitter()->Contains256bitAVX())
+    // This applies to 512bit AVX512 instructions as well.
+    if (call->IsPInvoke() && (call->gtCallType == CT_USER_FUNC) &&
+        (GetEmitter()->Contains256bitAVX() || GetEmitter()->Contains512bitAVX()))
```

Comment:
It might be worthwhile to collapse the `Contains256bitAVX` and `Contains512bitAVX` calls to a single method.

The intent behind them is really to check for usages of any YMM or greater size registers so we can do things like `vzeroupper` and avoid the penalty for code that is not VEX aware.

Notably we have a tracked work item to remove some usages of `vzeroupper` because we emit it almost everywhere.

In practice we should really only need to emit it as part of certain P/Invoke or Crossgen boundaries where we know we could be calling into something using the legacy encoding.

Reply:
That makes sense. I'll collapse both `contains256bitAVXInstruction` and `contains512bitAVXInstruction` plus all associated support functions to a single `contains256bitOrMoreAVXInstruction`

Reply:
Done